### ISSUE TITLE
AlbToFargate - unable to set new memory limit on TaskDefinition

### ISSUE THREAD

**Issue by yuflyud** [awslabs/aws-solutions-constructs#887]
By default the AlbToFargate construct creates a TaskDefinition resource with Cpu: "256" and Memory: "512". Changing those through the `AlbToFargateProps` builder's `fargateTaskDefinitionProps` (in Java) had no affect on `memory` property.

Also it is impossible to update the Container's `MemoryReservation` property. Calling the `containerDefinitionProps` requires to build a `ContainerDefinition` with a `TaskDefinition` reference, which is not present at this point.

### Reproduction Steps

Java example:
```
new AlbToFargate(this, "AlbToFargatePattern", new AlbToFargateProps.Builder().ecrRepositoryArn(
            "arn:aws:ecr:eu-central-1:<accountId>:repository/<repo>")
        .ecrImageVersion("latest")
        .publicApi(false)
        .logAlbAccessLogs(false)
        .fargateTaskDefinitionProps(TaskDefinitionProps.builder()
            .compatibility(Compatibility.FARGATE)
            .cpu("1024")
            .memoryMiB("2048")
            .build())
       .build());
```
Result:
```
AlbToFargatePatterntaskdef973C6679:
    Type: AWS::ECS::TaskDefinition
    Properties:
      ContainerDefinitions:
        - Essential: true
          Image:
            Fn::Join:
              - ""
              - - <accountId>.dkr.ecr.eu-central-1.
                - Ref: AWS::URLSuffix
                - /<repo>:latest
          MemoryReservation: 512
          Name: AlbToFargatePattern-container
          PortMappings:
            - ContainerPort: 8080
              Protocol: tcp
      Cpu: "1024"
      ExecutionRoleArn:
        Fn::GetAtt:
          - AlbToFargatePatterntaskdefExecutionRoleB00DE280
          - Arn
      Family: AwsCloudformationCdkStackAlbToFargatePatterntaskdefDBDED03F
      Memory: "512"
```

### Environment

  - **CDK CLI Version      :*2.59.0*
  - **CDK Framework Version:*2.59.0*
  - **AWS Solutions Constructs Version     :*2.30.0*
  - **OS                   :*macOS Monterey*
  - **Language             :*Java*

### Other

--- 

This is :bug: Bug Report

**Comment by biffgaut** [comment]
Thanks, we'll take a look

**Comment by georgebearden** [comment]
Hi @yuflyud - The issue has to do with the type you are passing into the `fargateTaskDefinitionProps`. If you change from using the `TaskDefinitionProps` to `FargateTaskDefinitionProps` and then set the corresponding `memoryLimitMib` property, your template will synthesize correctly, using both the overridden `cpu` and `memory` values you specify.

Regarding your other question on `MemoryReservation`, you should be able to add containers to the task definition after creating the construct. Along the lines of...

```typescript
const albToFargate = new AlbToFargate(stack, 'alb-to-fargate', {
    publicApi: false,
    ...
});

albToFargate.service.taskDefinition.addContainer('my-container', {
    image: EcsOptimizedImage,
    memoryReservationMiB: 512
})
```

**Comment by yuflyud** [comment]
Thank you @georgebearden . The FargateTaskDefinitionProps made it work as expected. Regarding changing the `MemoryReservation` property - I would like to set it for existing container, and not create a new one. In other case I will have a Task Definition with two containers.
I think this is not critical to have a configurable "MemoryReservation" property, as this is just a soft limit. But there is the same issue with setting a "Memory" property, which is sometimes really needed. Especially when having multiple containers within the same Task Definition, if I understand things correctly.

**Comment by georgebearden** [comment]
Hi @yuflyud - I see the distinction you are making regarding the code snippet I pasted above, where calling `addContainer` will actually create a second container, which is not what you want. Can you clarify something for me? the `MemoryLimit` and `MemoryReservation` properties exist on the `ContainerDefinition` itself, so if you have an existing container already, why can't those properties be set when the existing container is created elsewhere in your cdk code? It doesn't really make sense for us to change our API to expose these specific properties when they are already available on the `ContainerDefinition`.

**Comment by yuflyud** [comment]
Hi @georgebearden, I tried to supply my own ContainerDefinition object via existingContainerDefinitionObject method, but I get an error:
`Error: If an existing Service is indicated by supplying either existingFargateServiceObject or existingContainerDefinitionObject, then existingFargateServiceObject, existingContainerDefinitionObject, and existingVpc must all be provided
`
I know I can make it work, I just don't want to manually create a whole set of my own CF resources just to change a single property. I ended up creating the stack using ApplicationLoadBalancedFargateService construct and using AlbToFargate just as a reference implementation. This is of course my personal experience and I am not asking to fix anything for my specific use case, just sharing my thoughts. The initial issue has been resolved so feel free to close it. Thank you!
